### PR TITLE
Flask endpoints GPT API Call impl

### PR DESCRIPTION
**Pull request by willyg23**

### CODE REVIEW

**Comment by willyg23** [comment]
Description:
Adding flask endpoints and GPT implementation

Design choice explanation:
I decided to make the frontend request determine the AI model that the backend calls for. This is to reduce complexity and give us more flexibility in the long run. As if the backend were to determine the model that we’re using, at the minimum, we’d have to make a variable any time we want to use a different model. But if we make that decision the frontend request, that can change whenever we want, with no need to update any code.

So for the frontend: make a constant file, that contains constants. Such as the model’s ID. Wherever you make requests in your frontend code, use the variable defined in the constant file.
Why? Readability + organization.
 From a readability standpoint: 
This: 
      body: JSON.stringify({
        prompt: "How do cars in the movie Cars have kids?",
        model: CAR_ASSISTANT_MODEL_ID
      }),
    });

is much more readable than this:

      body: JSON.stringify({
        prompt: "How do cars in the movie Cars have kids?",
        model: g-gyCZHrDO1
      }),
    });


Because when you read the code, it’s obvious what CAR_ASSISTANT_MODEL_ID is. g-gyCZHrDO1 on the other hand, not so clear.

From a organization standpoint:
Constant variables are often used in multiple files throughout one codebase. Repeatedly defining them in each file is redundant and inefficient. Because if you have need to modify something, you have to go modify it in every file. Whereas with a constant file, you can just change that file.



The frontend POST request will need to be in this format: 
{
    "prompt": “text”,
    "model": “text”
}

File: backend/routes.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            messages=messages,
+            temperature=0
+        )
+    response_message = response.choices[0].message.content
+    print(response_message)
+    return response_message
+
+@app.route('/generate-gpt-response', methods=['POST'])
+def return_generated_response():
+    data = request.json
```
Comment (willyg23) [inline backend/routes.py:30]: collecting data from the frontend POST request

File: backend/routes.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    response_text = generate_gpt_response(data.get('prompt'), data.get('model'))
+    return jsonify({"response": response_text})
```
Comment (willyg23) [inline backend/routes.py:32]: return the gpt response back to the client (our app)

File: backend/routes.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+def generate_gpt_response(prompt, model):
+    messages = [
+            {"role": "system", "content": 'You answer question about cars. Most commonly, car maintenance'
+            },
+            {"role": "user", "content": prompt},
+        ]
+    response = client.chat.completions.create(
+            model=model,
+            messages=messages,
+            temperature=0
```
Comment (willyg23) [inline backend/routes.py:22]: temperature controls creativity or randomness of the response, on a scale from 0-2
0: most deterministic responses (focused and conservative).
1: balanced randomness (default).
2: Maximizes creativity and randomness (potentially less logical and more surprising responses).

I’ve noticed that higher temperature results in longer response times
When asking 0 and 2 temperature 


0 - 8.18 seconds
1 - 11.45 seconds
2 - 25.8 seconds

would be good to play around with some car maintenance questions that we know the proper answer to, and see if the temperature value affects the correctness of GPTs answers

File: backend/routes.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    api_key=os.environ.get("OPENAI_API_KEY"),
+)
+
+app = Flask(__name__)
+
+def generate_gpt_response(prompt, model):
+    messages = [
+            {"role": "system", "content": 'You answer question about cars. Most commonly, car maintenance'
+            },
+            {"role": "user", "content": prompt},
```
Comment (willyg23) [inline backend/routes.py:17]: defines that the message is coming from the user and the content of the message is their prompt

**Review by jduron24 — COMMENTED** [review]
👍

**Comment by jduron24** [comment]
swaf

**Review by jduron24 — APPROVED** [review]

**Pull request merged**